8238761: Asynchronous handshakes

robehn opened the pull request:
This patch implements asynchronous handshake, which changes how handshakes works by default. Asynchronous handshakes are target only executed, which they may never be executed. (target may block on socket for the rest of VM lifetime) Since we have several use-cases for them we can have many handshake pending. (should be very rare) To be able handle an arbitrary amount of handshakes this patch adds a per JavaThread queue and heap allocated HandshakeOperations. It's a singly linked list where you push/insert to the end and pop/get from the front. Inserts are done via CAS on first pointer, no lock needed. Pops are done while holding the per handshake state lock, and when working on the first pointer also CAS.

The thread grabbing the handshake state lock for a JavaThread will pop and execute all handshake operations matching the filter. The JavaThread itself uses no filter and any other thread uses the filter of everything except asynchronous handshakes. In this initial change-set there is no need to do any other filtering. If needed filtering can easily be exposed as a virtual method on the HandshakeClosure, but note that filtering causes handshake operation to be done out-order. Since the filter determins who execute the operation and not the invoked method, there is now only one method to call when handshaking one thread.

Some comments about the changes:
- HandshakeClosure uses ThreadClosure, since it neat to use the same closure for both alla JavThreads do and Handshake all threads. With heap allocating it cannot extends StackObj. I tested several ways to fix this, but those very much worse then this.

- I added a is_handshake_safe_for for checking if it's current thread is operating on itself or the handshaker of that thread.

- Simplified JVM TI with a JvmtiHandshakeClosure and also made them not needing a JavaThread when executing as a handshaker on a JavaThread, e.g. VM Thread can execute the handshake operation.

- Added WB testing method.

- Removed VM_HandshakeOneThread, the VM thread uses the same call path as direct handshakes did.

- Changed the handshake semaphores to mutex to be able to handle deadlocks with lock ranking.

- VM_HandshakeAllThreadsis still a VM operation, since we do support half of the threads being handshaked before a safepoint and half of them after, in many handshake all operations.

- ThreadInVMForHandshake do not need to do a fenced transistion since this is always a transistion from unsafe to unsafe.

- Added NoSafepointVerifyer, we are thinking about supporting safepoints inside handshake, but it's not needed at the moment. To make sure that gets well tested if added the NoSafepointVerifyer will raise eyebrows.

- Added ttyLocker::break_tty_lock_for_safepoint(os::current_thread_id()); due to lock rank.

- Added filtered queue and gtest for it.

Passes multiple t1-8 runs.
Been through some pre-reviwing.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8238761](https://bugs.openjdk.java.net/browse/JDK-8238761): Asynchronous handshakes


### Reviewers
 * [Patricio Chilano Mateo](https://openjdk.java.net/census#pchilanomate) (@pchilano - Committer) ⚠️ Review applies to cd784a751a3153939b9284898f370160124ca610
 * [Daniel D. Daugherty](https://openjdk.java.net/census#dcubed) (@dcubed-ojdk - **Reviewer**) ⚠️ Review applies to 469f8fc88092fb8bb7da43544c8d15c10de12986
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to cd784a751a3153939b9284898f370160124ca610
 * [Serguei Spitsyn](https://openjdk.java.net/census#sspitsyn) (@sspitsyn - **Reviewer**)


### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/151/head:pull/151`
`$ git checkout pull/151`

robehn commented:
> src/hotspot/share/prims/jvmtiThreadState.cpp
> src/hotspot/share/prims/jvmtiEnvBase.cpp

Removed double checks.

> src/hotspot/share/runtime/handshake.cpp
> ...
> I would just leave ProcessResult as an enum and log as before.

Reverted to plain enum and updated logs. (better?)

> src/hotspot/share/runtime/handshake.cpp
> 387: NoSafepointVerifier nsv;
> 388: process_self_inner();

I wanted a NSV to cover the process_self_inner method.
So I added a second one in suggested place:

>  NoSafepointVerifier nsv;
> _handshake_cl->do_thread(thread);
> }

> src/hotspot/share/runtime/interfaceSupport.inline.hpp
> 156: // Threads shouldn't block if they are in the middle of
> printing, but...
> 157: ttyLocker::break_tty_lock_for_safepoint(os::current_thread_id());
> What's the issue of having NoSafepointVerifier inside the handshake?

Sorry, the issue is the lock rank. Right now the semaphore hides this issue.

Please see commit 86b83d0.

pchilano commented on src/hotspot/share/runtime/handshake.cpp:
```diff
@@ -465,60 +443,87 @@ bool HandshakeState::possibly_can_process_handshake() {
-HandshakeState::ProcessResult HandshakeState::try_process(HandshakeOperation* op) {
-  bool is_direct = op->is_direct();
+const char* executioner_name(Thread* current_thread, Thread* handshakee, bool current_is_requester) {
+  if (current_thread == handshakee) return "self(JavaThread)";
```
I think we can remove this line since executioner_name() is only called by the handshaker.

robehn replied:
Yes, fixing.

pchilano commented on src/hotspot/share/runtime/handshake.cpp:
```diff
@@ -465,60 +443,87 @@ bool HandshakeState::possibly_can_process_handshake() {
+      assert(SafepointMechanism::local_poll_armed(_handshakee), "Must be");
+      assert(op->_target == NULL || _handshakee == op->_target, "Wrong thread");
+      log_trace(handshake)("Processing handshake " INTPTR_FORMAT " by %s", p2i(op),
+                           executioner_name(current_thread, _handshakee, op == match_op));
```
With the above change we could even avoid factoring the code into executioner_name() and just do: 
log_trace(handshake)("Processing handshake " INTPTR_FORMAT " by %s%s", p2i(op),
                                         op == match_op ? "handshaker" : "cooperative", 
                                         current_thread->is_VM_thread() ? "(VM Thread)" : "(JavaThread)");

robehn replied:
I added a second log line where I use that function also:
  log_trace(handshake)("Thread %s(" INTPTR_FORMAT ") executed %d ops for JavaThread: " INTPTR_FORMAT " %s target op: " INTPTR_FORMAT,
                       executioner_name(current_thread, _handshakee, pr_ret == HandshakeState::_succeed),
                       p2i(current_thread), executed, p2i(_handshakee),
                       pr_ret == HandshakeState::_succeed ? "including" : "excluding", p2i(match_op));

robehn replied:
Push commit 469f8fc, please have a look.

pchilano commented on src/hotspot/share/prims/jvmtiEnvBase.cpp:
```diff
@@ -904,7 +906,7 @@ JvmtiEnvBase::get_frame_location(JavaThread *java_thread, jint depth,
 #endif
   Thread* current_thread = Thread::current();
   assert(current_thread == java_thread ||
```
One extra check here.

robehn replied:
Fixing

pchilano commented on src/hotspot/share/prims/jvmtiEnvBase.cpp:
```diff
@@ -1161,7 +1163,7 @@ MultipleStackTracesCollector::fill_frames(jthread jt, JavaThread *thr, oop threa
   Thread *current_thread = Thread::current();
   assert(current_thread == thr ||
```
One extra check here.

robehn replied:
Fixing

pchilano reviewed (commented):
Changes look good, thanks for fixing! I added some comments on the changes.

pchilano commented:
Update looks good, thanks Robbin!

dcubed-ojdk commented on src/hotspot/share/prims/jvmtiEnvBase.cpp:
```diff
@@ -649,10 +649,9 @@ JvmtiEnvBase::count_locked_objects(JavaThread *java_thread, Handle hobj) {
-         current_jt == java_thread->active_handshaker(),
-         "call by myself or at direct handshake");
+  Thread *current_thread = Thread::current();
+  assert(java_thread->is_handshake_safe_for(current_thread),
```
I like how this assert reads now!

robehn replied:
Great!

dcubed-ojdk commented on src/hotspot/share/prims/jvmtiEventController.cpp:
```diff
@@ -333,11 +333,10 @@ void JvmtiEventControllerPrivate::enter_interp_only_mode(JvmtiThreadState *state
     hs.do_thread(target);
   } else {
-    bool executed = Handshake::execute_direct(&hs, target);
-    guarantee(executed, "Direct handshake failed. Target thread is not alive?");
```
This guarantee() that the handshake has executed doesn't have an
equivalent in the rewritten code. Should there be some way of verifying
this condition from this location?

robehn replied:
Fixed

dcubed-ojdk commented on src/hotspot/share/runtime/handshake.hpp:
```diff
@@ -39,74 +40,75 @@ class JavaThread;
-    _state_busy,
-    _success,
-    _number_states
+      _no_operation = 0,
+      _not_safe,
+      _claim_failed,
+      _processed,
+      _succeed,
+      _number_states
```
Why are these indented by 4 spaces instead of 2 spaces?

robehn replied:
Fixed

dcubed-ojdk commented on src/hotspot/share/runtime/handshake.cpp:
```diff
@@ -30,37 +30,44 @@
+  AsyncHandshakeOperation(AsynchHandshakeClosure* cl, JavaThread* target, jlong start_ns)
+    : HandshakeOperation(cl, target), _start_time_ns(start_ns) {}
+  virtual ~AsyncHandshakeOperation() { delete _handshake_cl; };
+  jlong start_time()                 { return _start_time_ns; }
```
Should this be 'const'? Ignore it if it would fan out too much.

robehn replied:
Fixed

dcubed-ojdk commented on src/hotspot/share/runtime/handshake.cpp:
```diff
@@ -373,39 +335,61 @@ bool Handshake::execute_direct(HandshakeClosure* thread_cl, JavaThread* target)
+  if (tlh.includes(target)) {
+    target->handshake_state()->add_operation(op);
+  } else {
+    log_handshake_info(start_time_ns, op->name(), 0, 0, "(thread dead)");
```
It might be useful to also log the 'target' thread value here so:

    .... (thread=<ptr-value> is dead)"

Might be something like this:

log_handshake_info(start_time_ns, op->name(), 0, 0, "(thread=" INTPTR_FORMAT " is dead)", p2i(target));

Although you'd probably have to form the string in a buffer and then pass it
to the log_handshake_info() call... sigh...

robehn replied:
Fixed (via buffert)

dcubed-ojdk commented on src/hotspot/share/runtime/handshake.cpp:
```diff
@@ -465,60 +443,78 @@ bool HandshakeState::possibly_can_process_handshake() {
   }
-  if (has_specific_operation(is_direct)){
-    return true;
+  // Operations are added without lock and then the poll is armed.
```
s/without lock/lock free/

robehn replied:
Fixed

dcubed-ojdk commented on src/hotspot/share/runtime/handshake.cpp:
```diff
@@ -465,60 +443,78 @@ bool HandshakeState::possibly_can_process_handshake() {
-  }
-
-  // If we own the semaphore at this point and while owning the semaphore
+  // If we own the mutex at this point and while owning the mutex
```
grammar - s/owning the mutex/owning the mutex we/

robehn replied:
Fixed

dcubed-ojdk commented on src/hotspot/share/runtime/interfaceSupport.inline.hpp:
```diff
@@ -156,6 +152,9 @@ class ThreadInVMForHandshake : public ThreadStateTransition {
     thread->set_thread_state(_thread_in_vm);
+
+    // Threads shouldn't block if they are in the middle of printing, but...
+    ttyLocker::break_tty_lock_for_safepoint(os::current_thread_id());
```
Can you explain why you had to add this?
Did something show up in testing?

robehn replied:
The tty lock allow vm thread to block and it is a no safepoint check lock.
Meaning that you are not allowed to safepoint/handshake poll while holding it.
Also if someone is holding it, you can't print on the tty inside a safepoint/handshake.

The tty lock is also lower rank than Threads_lock (we used to block on that in safepoint) and lower rank than the handshake state lock.

Compiler have code which takes the ttyLocker and prints on tty.
Some of the information that it prints passes code which can safepoint/handshake.
The way safepoint have gotten around this is by unlocking tty lock before a safepoint.
Now that handshake use a mutex instead of semaphore, we thus get an error also.
(even if you now don't get an error we could potentially deadlock, or at least we are very deep water).

Until compiler is fixed to UL we need this unlocking of the tty lock.

dcubed-ojdk commented on src/hotspot/share/runtime/thread.cpp:
```diff
@@ -478,12 +478,13 @@ Thread::~Thread() {
-         !thread->as_Java_thread()->on_thread_list() ||
+  assert(!thread->is_Java_thread() ||
+         ((JavaThread *) thread)->is_handshake_safe_for(Thread::current()) ||
+         !((JavaThread *) thread)->on_thread_list() ||
```
Should use "thread->as_Java_thread()" instead of the cast here (2 places).

robehn replied:
Fixed

dcubed-ojdk commented on src/hotspot/share/runtime/thread.hpp:
```diff
@@ -1353,24 +1353,11 @@ class JavaThread: public Thread {
-    return _handshake.active_handshaker();
+  bool is_handshake_safe_for(Thread* th) const {
+    return _handshake.active_handshaker() == th ||
+           this == th;
```
I _think_ L1359-60 will fit on one line...

robehn replied:
Fixed

dcubed-ojdk commented on src/hotspot/share/utilities/filterQueue.inline.hpp:
```diff
@@ -0,0 +1,114 @@
+void FilterQueue<E>::add(E data) {
+  FilterQueueNode* head;
+  FilterQueueNode* insnode = new FilterQueueNode(data);
+  SpinYield yield(SpinYield::default_spin_limit * 10); // Very unlikely with mutiple failed CAS.
```
Typo - s/mutiple/multiple/

dcubed-ojdk commented on src/hotspot/share/utilities/filterQueue.inline.hpp:
```diff
@@ -0,0 +1,114 @@
+  if (cur == NULL) {
+    return (E)NULL;
+  }
+  SpinYield yield(SpinYield::default_spin_limit * 10); // Very unlikely with mutiple failed CAS.
```
typo - s/mutiple/multiple/

robehn replied:
Fixed

dcubed-ojdk commented on src/hotspot/share/prims/whitebox.cpp:
```diff
@@ -2025,6 +2025,29 @@ WB_ENTRY(jint, WB_HandshakeWalkStack(JNIEnv* env, jobject wb, jobject thread_han
+  class TraceSelfClosure : public AsynchHandshakeClosure {
+    void do_thread(Thread* th) {
+      assert(th->is_Java_thread(), "sanity");
+      JavaThread* jt = (JavaThread*)th;
```
Can whitebox.cpp code use the new as_Java_thread() call?

robehn replied:
Fixed :)

dcubed-ojdk commented on src/hotspot/share/prims/whitebox.cpp:
```diff
@@ -2025,6 +2025,29 @@ WB_ENTRY(jint, WB_HandshakeWalkStack(JNIEnv* env, jobject wb, jobject thread_han
+    void do_thread(Thread* th) {
+      assert(th->is_Java_thread(), "sanity");
+      JavaThread* jt = (JavaThread*)th;
+      ResourceMark rm;
```
It also might be interesting to print the "current thread" info here so
that someone looking at the test output knows which thread handled
the handshake (the target or a surrogate).

robehn replied:
It's always the thread it self, added asserts to make this clear.

dcubed-ojdk commented on src/hotspot/share/runtime/handshake.hpp:
```diff
@@ -39,74 +40,75 @@ class JavaThread;
-class HandshakeClosure : public ThreadClosure {
+// a single target/direct handshake or not, by the JavaThread that requested the
+// handshake or the VMThread respectively.
+class HandshakeClosure : public ThreadClosure, public CHeapObj<mtThread> {
```
Just to be clear. You haven't added support for a handshake that
must only be executed by the target thread yet, right? That's
future work, if I remember correctly...

robehn replied:
Yes as David says. The code that make sure this is case is:
static bool processor_filter(HandshakeOperation* op) {
  return !op->is_asynch();
}
Which is used e.g:
  return _queue.contains(processor_filter);

While process_self_inner() uses no filter.

dcubed-ojdk reviewed (approved):
Thumbs up. I don't think I have anything that is in the must fix category.

dholmes-ora commented on src/hotspot/share/utilities/filterQueue.inline.hpp:
```diff
@@ -0,0 +1,114 @@
+  return false;
+}
+
+// MT-Unsafe, external serialization needed.
```
So IIUC this queue supports multiple concurrent add()s, but has to be restricted to a single pop() at a time (although that is allowed to execute concurrently with the add()s) - correct?

robehn replied:
Yes

dholmes-ora commented on test/hotspot/jtreg/runtime/handshake/AsyncHandshakeWalkStackTest.java:
```diff
@@ -0,0 +1,111 @@
+ * @library /testlibrary /test/lib
+ * @build AsyncHandshakeWalkStackTest
+ * @run driver ClassFileInstaller sun.hotspot.WhiteBox
+ *                              sun.hotspot.WhiteBox$WhiteBoxPermission
```
This is not needed as ClassFileInstaller already handles WhiteBox's nested classes.

robehn replied:
Fixed

dholmes-ora commented on test/hotspot/jtreg/runtime/handshake/MixedHandshakeWalkStackTest.java:
```diff
@@ -0,0 +1,67 @@
+ * @library /testlibrary /test/lib
+ * @build MixedHandshakeWalkStackTest
+ * @run driver ClassFileInstaller sun.hotspot.WhiteBox
+ *                              sun.hotspot.WhiteBox$WhiteBoxPermission
```
Not needed - see earlier comment.

robehn replied:
Fixed

dholmes-ora commented on test/hotspot/jtreg/runtime/handshake/MixedHandshakeWalkStackTest.java:
```diff
@@ -0,0 +1,67 @@
+import sun.hotspot.WhiteBox;
+
+public class MixedHandshakeWalkStackTest {
+    public static Thread tthreads[];
```
why tthreads? It looks like a typo :)

robehn replied:
Test threads, changed to testThreads.

dholmes-ora reviewed (commented):
Hi Robbin,

There is still a lack of motivation for this feature in the JBS issue. What kind of handshakes need to be asynchronous? Any async operation implies that the requester doesn't care about when or even if, the operation gets executed - they are by definition fire-and-forget actions. So what are the usecases being envisaged here?

Many of the changes included here seem unrelated to, and not reliant on, async handshakes, and could be factored out to simplify the review and allow focus on the actual async handshake part e.g. the JVM TI cleanups seem they could be mostly standalone.

Specific comments below. A general concern I have is where the current thread is no longer guaranteed to be a JavaThread (which is a step in the wrong direction in relation to some of the cleanups I have planned!) and I can't see why this would be changing.

Thanks.

robehn commented on src/hotspot/share/prims/jvmtiEnvBase.cpp:
```diff
@@ -687,15 +686,19 @@ JvmtiEnvBase::get_current_contended_monitor(JavaThread *calling_thread, JavaThre
                                  GrowableArray<jvmtiMonitorStackDepthInfo*> *owned_monitors_list) {
+  // Note:
+  // calling_thread is the thread that requested the list of monitors for java_thread.
+  // java_thread is thread owning the monitors.
```
Fixed

robehn commented on src/hotspot/share/prims/jvmtiEnvBase.cpp:
```diff
@@ -687,15 +686,19 @@ JvmtiEnvBase::get_current_contended_monitor(JavaThread *calling_thread, JavaThre
+  // Note:
+  // calling_thread is the thread that requested the list of monitors for java_thread.
+  // java_thread is thread owning the monitors.
+  // current_thread is thread executint this code, can be a non-JavaThread (e.g. VM Thread).
```
Fixed

robehn commented on src/hotspot/share/prims/jvmtiEnvBase.cpp:
```diff
@@ -687,15 +686,19 @@ JvmtiEnvBase::get_current_contended_monitor(JavaThread *calling_thread, JavaThre
+  // calling_thread is the thread that requested the list of monitors for java_thread.
+  // java_thread is thread owning the monitors.
+  // current_thread is thread executint this code, can be a non-JavaThread (e.g. VM Thread).
+  // And they all maybe different threads.
```
Fixed

robehn commented on src/hotspot/share/prims/jvmtiEnvBase.hpp:
```diff
@@ -336,58 +336,58 @@ class JvmtiEnvIterator : public StackObj {
 
+class JvmtiHandshakeClosure : public HandshakeClosure {
+ protected:
+  jvmtiError _result;
```
Thanks

robehn commented:
> There is still a lack of motivation for this feature in the JBS issue. What kind of handshakes need to be asynchronous? Any async operation implies that the requester doesn't care about when or even if, the operation gets executed - they are by definition fire-and-forget actions. So what are the usecases being envisaged here?
> 

Added info in JBS issue:
https://bugs.openjdk.java.net/browse/JDK-8238761

All uses-cases of _suspend_flag as initial targets.
But some of them require more bits.

> Many of the changes included here seem unrelated to, and not reliant on, async handshakes, and could be factored out to simplify the review and allow focus on the actual async handshake part e.g. the JVM TI cleanups seem they could be mostly standalone.
> 

Since I kept rebasing this doing somethings I did somethings to simplify the rebasing.
I guess you are talking about the JvmtiHandshakeClosure?

> Specific comments below. A general concern I have is where the current thread is no longer guaranteed to be a JavaThread (which is a step in the wrong direction in relation to some of the cleanups I have planned!) and I can't see why this would be changing.
> 

If the VM thread emits a "handshake all" it will continuously loop the JavaThreads until op is completed.
We do not keep track which JavaThread have done which handshake.
The VM thread will execute all handshakes it finds on every JavaThread's queue.
If someone else adds a handshake to one of the JavaThreads the VM thread might execute it.

I did not see any issues while looking at the code or in testing doing this.
Many of the handshakes used to be safepoints, always executed by VM thread.
We changed that to VM thread or JavaThread it self, and we now extend that to any JavaThread.

Some of the JVM TI handshakes are a bit different, but since they must proper allocate resource in target JavaThread and not in current JavaThread, there is no issue executing the code with a non-JavaThread.

At the moment we have no dependencies on that the 'driver' is a JavaThread for any of the handshakes.
We can easily set a per Handshake typ filter (slight changes to Handshake Closure and filter function) and choose to only executed the handshake with target self and requester/any JavaThread/only VM thread.

So if we think JVM TI handshakes should only be executed by requester or target it's an easy fix.
If you think the default is wrong, it's also an easy change.

(For others following there also a planned investigation on requester only executed handshake, which is not as easy)

> Thanks.

#

robehn commented on src/hotspot/share/prims/jvmtiEnvThreadState.cpp:
```diff
@@ -314,11 +314,10 @@ void JvmtiEnvThreadState::reset_current_location(jvmtiEvent event_type, bool ena
       } else {
-        bool executed = Handshake::execute_direct(&op, _thread);
-        guarantee(executed, "Direct handshake failed. Target thread is not alive?");
+        Handshake::execute(&op, _thread);
```
Fixed

robehn commented on src/hotspot/share/prims/jvmtiEventController.cpp:
```diff
@@ -333,11 +333,10 @@ void JvmtiEventControllerPrivate::enter_interp_only_mode(JvmtiThreadState *state
   } else {
-    bool executed = Handshake::execute_direct(&hs, target);
-    guarantee(executed, "Direct handshake failed. Target thread is not alive?");
+    Handshake::execute(&hs, target);
```
Fixed

robehn commented on src/hotspot/share/prims/jvmtiEnvBase.cpp:
```diff
@@ -687,15 +686,19 @@ JvmtiEnvBase::get_current_contended_monitor(JavaThread *calling_thread, JavaThre
+  // Note:
+  // calling_thread is the thread that requested the list of monitors for java_thread.
+  // java_thread is thread owning the monitors.
+  // current_thread is thread executint this code, can be a non-JavaThread (e.g. VM Thread).
```
The assumption that it must be an JavaThread was due to code such as "JavaThread::current();", any Thread can execute this.
The default the first thread grabbing the handshake state lock for the targeted thread will execute all handshake operation it is allowed to. This might be the VM thread.

robehn commented on src/hotspot/share/runtime/handshake.cpp:
```diff
@@ -186,85 +192,42 @@ void VM_Handshake::handle_timeout() {
 
     log_trace(handshake)("Threads signaled, begin processing blocked threads by VMThread");
     HandshakeSpinYield hsy(start_time_ns);
+    int executed_by_driver = 0;
```
The driver is VM thread or a JavaThread.

robehn commented on src/hotspot/share/runtime/handshake.cpp:
```diff
@@ -327,37 +290,36 @@ void HandshakeOperation::do_handshake(JavaThread* thread) {
   VMThread::execute(&handshake);
 }
 
-bool Handshake::execute(HandshakeClosure* thread_cl, JavaThread* target) {
```
As I said in PR request:
"Removed VM_HandshakeOneThread, the VM thread uses the same call path as direct handshakes did."
They are not VM ops, and default we do not promise who will execute it.
Just that it will be executed on thread X.

robehn commented on src/hotspot/share/runtime/handshake.cpp:
```diff
@@ -327,37 +290,36 @@ void HandshakeOperation::do_handshake(JavaThread* thread) {
   }
 
-  HandshakeState::ProcessResult pr =  HandshakeState::_no_operation;
+  int executed_by_driver = 0;
```
The thread executing this handshake operation, what the current thread is doesn't matter.
You can't use current threads resources or be dependent otherwise on it.

Exception being locking issues in JVM TI, where we are dependent that requester have locked JVM TI state lock for us, but we are not dependent that the current thread is the owner. So checking that the lock is held by requester doesn't matter for how is the 'driver'.

robehn commented on src/hotspot/share/runtime/handshake.cpp:
```diff
@@ -465,60 +443,78 @@ bool HandshakeState::possibly_can_process_handshake() {
+  // Operations are added without lock and then the poll is armed.
+  // If all handshake operations for the handshakee are finished and someone
+  // just adds an operation we may see it here. But if the handshakee is not
+  // armed yet it is not safe to procced.
```
Fixed

robehn commented on src/hotspot/share/runtime/handshake.hpp:
```diff
@@ -39,74 +40,75 @@ class JavaThread;
-// or the VMThread respectively.
-class HandshakeClosure : public ThreadClosure {
+// a single target/direct handshake or not, by the JavaThread that requested the
+// handshake or the VMThread respectively.
```
The concept of direct handshake do not exist in that way. (but can easily be implemented using the filter)
You have operation that you need to be executed on a JavaThread, you add that to that JavaThread.
Any thread ("driver") that succeed to claim that JavaThreads handshake state (lock and that JavaThread is safe) procced to execute from that handshake queue until empty (empty according to applied filter on queue).

robehn commented on src/hotspot/share/runtime/handshake.hpp:
```diff
@@ -39,74 +40,75 @@ class JavaThread;
+      _not_safe,
+      _claim_failed,
+      _processed,
+      _succeed,
```
Fixed

robehn commented on src/hotspot/share/runtime/interfaceSupport.inline.hpp:
```diff
@@ -132,10 +132,6 @@ class ThreadInVMForHandshake : public ThreadStateTransition {
     // This can be invoked from transition states and must return to the original state properly
     assert(_thread->thread_state() == _thread_in_vm, "should only call when leaving VM after handshake");
-    // Change to transition state and ensure it is seen by the VM thread.
-    _thread->set_thread_state_fence(_thread_in_vm_trans);
```
You can never call SafepointMechanism::process_if_requested() while in any 'safe-state' (as determine by safepoint/handshake safe routines).
We thus know we are transitioning from an unsafe state to an unsafe state, which unsafe state might be seen by e.g. VM thread we care little about.

robehn commented on src/hotspot/share/runtime/thread.cpp:
```diff
@@ -478,12 +478,13 @@ Thread::~Thread() {
 // A JavaThread is considered "dangling" if it is not the current
-// thread, has been added the Threads list, the system is not at a
-// safepoint and the Thread is not "protected".
+// thread, his not handshaking with current thread, as been added the Threads
```
Fixed

robehn commented on src/hotspot/share/runtime/thread.cpp:
```diff
@@ -847,7 +848,7 @@ bool JavaThread::wait_for_ext_suspend_completion(int retries, int delay,
 bool
 JavaThread::is_thread_fully_suspended(bool wait_for_suspend, uint32_t *bits) {
-  if (this != JavaThread::current()) {
+  if (this != Thread::current()) {
```
is_thread_fully_suspended() is used in JVM TI handshake and this change-set allow VM thread to execute them.
This method have no dependency that caller is a JavaThread.

robehn commented on src/hotspot/share/runtime/thread.cpp:
```diff
@@ -2436,7 +2436,9 @@ void JavaThread::handle_special_runtime_exit_condition(bool check_asyncs) {
-  assert(Thread::current()->is_VM_thread() || Thread::current() == this, "should be in the vm thread");
+  assert(Thread::current()->is_VM_thread() ||
+         is_handshake_safe_for(Thread::current()),
+         "should be in the vm thread, self or handshakee");
```
send_thread_stop() is only called from a handshake operation.
Fixed.

robehn commented on src/hotspot/share/utilities/filterQueue.hpp:
```diff
@@ -0,0 +1,75 @@
+  }
+
+  // MT-safe
+  // Since pops and adds are allowed while we add, we do not know if _first is same even if it's the same address.
```
Removed comment.
Ok. Since it's same notation as posix uses I thought was clear that any method marked MT-Unsafe is not safe call to from a multithreaded program.

Suggestion on how to separate the MT-safe from MT-usafe methods ?

robehn commented on src/hotspot/share/utilities/filterQueue.inline.hpp:
```diff
@@ -0,0 +1,114 @@
+    if (Atomic::cmpxchg(&_first, head, insnode) == head) {
+      break;
+    }
+    yield.wait();
```
No and no.
It was added to guard against any pathologically case, such as the gtest stress-tests.

coleenp reviewed (changes requested):
Looks mostly good to me!

coleenp commented on test/hotspot/jtreg/runtime/handshake/HandshakeDirectTest.java:
```diff
@@ -39,7 +39,7 @@
 public class HandshakeDirectTest  implements Runnable {
     static final int WORKING_THREADS = 32;
-    static final int DIRECT_HANDSHAKES_MARK = 50000;
+    static final int DIRECT_HANDSHAKES_MARK = 500000;
```
Could this timeout?

robehn replied:
I have seen no signs of it and have been through several mach5 runs with no issues.

dholmes-ora commented:
Hi Robbin,

I've gone back to refresh myself on the previous discussions and (internal) design walk-throughs to get a better sense of these changes. Really the "asynchronous handshake" aspect is only a small part of this. The fundamental change here is that handshakes are now maintained via a per-thread queue, and those handshake operations can, in the general case, be executed by any of the target thread, the requestor (active_handshaker) thread or the VMThread. Hence the removal of the various "JavaThread::current()" assumptions. Unless constrained otherwise, any handshake operation may be executed by the VMThread so we have to take extra care to ensure the code is written to allow this. I'm a little concerned that our detour into direct-handshakes actually lulled us into a false sense of security knowing that an operation would always execute in a JavaThread, and we have now reverted that and allowed the VMThread back in. I understand why, but the change in direction here caught me by surprise (as I had forgotten the bigger picture). It may not always be obvious that the transitive closure of the code from an operation can be safely executed by a non-JavaThread.

Then on top of this generalized queuing mechanism there is a filter which allows some control over which thread may perform a given operation - at the moment the only filter isolates "async" operations which only the target thread can execute.

In addition another nuance is that when processing a given thread's handshake operation queue, different threads have different criteria for when to stop processing the queue:
- the target thread will drain the queue completely
- the VMThread will drain the queue of all "non-async" operations**
- the initiator for a "direct" operation will drain the queue up to, and including, the synchronous operation they submitted
- the initiator for an "async" operation will not process any operation themselves and will simply add to the queue and then continue on their way (hence the "asynchronous")

** I do have some concerns about latency impact on the VMThread if it is used to execute operations that didn't need to be executed by the VMThread!

I remain concerned about the terminology conflation that happens around "async handshakes". There are two aspects that need to be separated:
- the behaviour of the thread initiating the handshake operation; and
- which thread can execute the handshake operation

When a thread initiates a handshake operation and waits until that operation is complete (regardless of which thread performed it, or whether the initiator processed any other operations) that is a synchronous handshake operation.
When a thread initiates a handshake operation and does not wait for the operation to complete (it does the target->queue()->add(op); and continues on its way) that is an asynchronous handshake operation.

The question of whether the operation must be executed by the target thread is orthogonal to whether the operation was submitted as a synchronous or asynchronous operation.

So I have problem when you say that an asynchronous handshake operation is one that must be executed by the target thread, as this is not the right characterisation at all. It is okay to constrain things such that an async operation is always executed by the target, but that is not what makes it an async operation. In the general case there is no reason why an async operation might not be executed by the VMThread, or some other JavaThread performing a synchronous operation on the same target.

I will go back through the actual handshake code to see if there are specific things I would like to see changed, but that will have to wait until tomorrow.

Thanks,
David

robehn commented:
Hi David, you are correct and you did a fine job summarizing this, thanks!

> I will go back through the actual handshake code to see if there are specific things I would like to see changed, but that will have to wait until tomorrow.

Great thanks!

> 
> Thanks,
> David

#

robehn commented on src/hotspot/share/runtime/handshake.hpp:
```diff
@@ -39,74 +40,75 @@ class JavaThread;
   virtual void do_thread(Thread* thread) = 0;
 };
 
+class AsynchHandshakeClosure : public HandshakeClosure {
```
Fixed

robehn commented on src/hotspot/share/runtime/handshake.hpp:
```diff
@@ -39,74 +40,75 @@ class JavaThread;
-  bool _thread_in_process_handshake;
+  FilterQueue<HandshakeOperation*> _queue;
+  Mutex _lock;
+  Thread* _active_handshaker;
```
The order of members matter since C++ initialize them in declared order.
My opinion when changing this was that it was easier to read when passing the only argument to the first member being initialized, thus _handshakee must be first member.

But I should init _active_handshaker in constructor, so added that and lined-up.

So before I do any such change please reflect over how the constructor will look like.

coleenp replied:
I don't understand, you'd have to rearrange the initializers in the constructor too, but I don't see any order dependance.  Moving over _lock  helps, so this is fine.

robehn replied:
You want a cosmetic change in the member declaration.
I'm saying the constructor will look worse.

Im asking if you want to trade a worse constructor for that?
(all here is extremely subjective :) )

coleenp replied:
It is subjective.  What you have is fine.

robehn commented on src/hotspot/share/runtime/handshake.cpp:
```diff
@@ -373,39 +337,61 @@ bool Handshake::execute_direct(HandshakeClosure* thread_cl, JavaThread* target)
+  ThreadInVMForHandshake tivm(_handshakee);
+  {
+    NoSafepointVerifier nsv;
+    process_self_inner();
```
We might remove ThreadInVMForHandshake and NoSafepointVerifier in the future.
ThreadInVMForHandshake is needed while we have the _suspend_flag and we have two different request to safepoint inside handshake due java heap allocation inside handshakes.
Having them outside like now it's clearer if we do such changes, so I'll keep it this way.

robehn commented:
@coleenp I think you placed your comment:
"The "driver" concept is odd. Should it really be caller? Like the thread that called VMHandshake?"
On the commit or somewhere else, not a review comment.
Anyhow I can reply to it here.

The answer is caller makes me think of requester, but we don't know if it's the requester executing this.
Therefore I referred to the thread executing handshakes on the target/handshakee as driver.

coleenp commented on src/hotspot/share/prims/whitebox.cpp:
```diff
@@ -2025,6 +2025,32 @@ WB_ENTRY(jint, WB_HandshakeWalkStack(JNIEnv* env, jobject wb, jobject thread_han
+  if (thread_oop != NULL) {
+    JavaThread* target = java_lang_Thread::thread(thread_oop);
+    TraceSelfClosure* tsc = new TraceSelfClosure(target);
+    Handshake::execute(tsc, target);
```
I know it's a whitebox test, but should this delete TraceSelfClosure sometime?

robehn replied:
Please have a look here, we delete it if we fail to install it:
https://github.com/openjdk/jdk/blob/cd784a751a3153939b9284898f370160124ca610/src/hotspot/share/runtime/handshake.cpp#L352
Here we delete it after we processed it:
https://github.com/openjdk/jdk/blob/cd784a751a3153939b9284898f370160124ca610/src/hotspot/share/runtime/handshake.cpp#L419

Since this as async op, the requester can't safely delete it.

coleenp replied:
Ok, thanks for pointing it out.

coleenp reviewed (changes requested):
Just one comment that needs an answer if you can find it :)

coleenp reviewed (approved):
Ok, all questions addressed.  Looks good!

dholmes-ora commented on src/hotspot/share/runtime/thread.hpp:
```diff
@@ -1353,24 +1353,12 @@ class JavaThread: public Thread {
 
-  Thread* active_handshaker() const {
-    return _handshake.active_handshaker();
+  // A JavaThread can always safely operate on it self and other threads
```
s/it self/itself/

robehn replied:
Fixed

dholmes-ora commented on src/hotspot/share/runtime/thread.hpp:
```diff
@@ -1353,24 +1353,12 @@ class JavaThread: public Thread {
-  Thread* active_handshaker() const {
-    return _handshake.active_handshaker();
+  // A JavaThread can always safely operate on it self and other threads
+  // can do it safely it if they are the active handshaker.
```
s/it if/if/

robehn replied:
Fixed

dholmes-ora commented on src/hotspot/share/utilities/filterQueue.hpp:
```diff
@@ -0,0 +1,73 @@
+#include "runtime/atomic.hpp"
+
+template <class E>
+class FilterQueue {
```
A brief description of the class would be good. It is basically a FIFO queue but with the ability to skip nodes that match a given "filter" criteria.

robehn replied:
Added

dholmes-ora commented on src/hotspot/share/utilities/filterQueue.hpp:
```diff
@@ -0,0 +1,73 @@
+template <class E>
+class FilterQueue {
+ private:
+  class FilterQueueNode : public CHeapObj<mtInternal> {
```
The Filter in FilterQueueNode is redundant given this is a nested type. Node would suffice.

robehn replied:
Fixed

dholmes-ora commented on src/hotspot/share/utilities/filterQueue.hpp:
```diff
@@ -0,0 +1,73 @@
+    return load_first() == NULL;
+  }
+
+  // MT-safe
```
Not sure where our previous discussion is on this but these posix-style MT-safe labels don't really halp for this kind of abstract data type API. Please briefly explain the thread-safety properties of add and pop.

robehn replied:
Added comment

dholmes-ora commented on src/hotspot/share/utilities/filterQueue.hpp:
```diff
@@ -0,0 +1,73 @@
+  }
+
+  // MT-safe
+  void add(E data);
```
It would be more regular naming to use add/remove or push/pop rather than add/pop.

robehn replied:
Change to push.

dholmes-ora reviewed (changes requested):
No significant comments. All my concerns relate to naming and terminology, where I think there is scope for quite a bit of tidying up. Thanks.

robehn commented on src/hotspot/share/runtime/handshake.hpp:
```diff
@@ -39,74 +40,75 @@ class JavaThread;
-class HandshakeClosure : public ThreadClosure {
+// a single target/direct handshake or not, by the JavaThread that requested the
+// handshake or the VMThread respectively.
+class HandshakeClosure : public ThreadClosure, public CHeapObj<mtThread> {
```
There is other 'sane' way to 'implement an interface' and being heap allocated.
We have a few like:
class DefaultICProtectionBehaviour: public CompiledICProtectionBehaviour, public CHeapObj<mtInternal>
class AOTCompiledMethod : public CompiledMethod, public CHeapObj<mtCode>
template <MEMFLAGS F> class TaskQueueSetSuperImpl: public CHeapObj<F>, public TaskQueueSetSuper

So the 'rule' about multiple inheritance is in conflict with not using new/delete operator without using CHeapObj.

robehn commented on src/hotspot/share/runtime/handshake.hpp:
```diff
@@ -39,74 +40,75 @@ class JavaThread;
-  }
+  virtual ~HandshakeClosure() {}
+  const char* name() const    { return _name; }
+  virtual bool is_asynch()    { return false; };
```
Missed this one, fixing!

robehn commented on src/hotspot/share/runtime/handshake.hpp:
```diff
@@ -39,74 +40,75 @@ class JavaThread;
-  bool _thread_in_process_handshake;
+  FilterQueue<HandshakeOperation*> _queue;
+  Mutex   _lock;
+  Thread* _active_handshaker;
```
Fixed

robehn commented on src/hotspot/share/runtime/handshake.hpp:
```diff
@@ -39,74 +40,75 @@ class JavaThread;
-  void set_handshakee(JavaThread* thread) { _handshakee = thread; }
+  void add_operation(HandshakeOperation* op);
+  HandshakeOperation* pop_for_self();
+  HandshakeOperation* pop_for_processor();
```
Fixing

robehn commented on src/hotspot/share/runtime/handshake.hpp:
```diff
@@ -39,74 +40,75 @@ class JavaThread;
-      AutoModifyRestore<bool> temporarily(_thread_in_process_handshake, true);
-      process_self_inner();
-    }
+  bool block_for_operation() {
```
Fixed

robehn commented on src/hotspot/share/runtime/handshake.hpp:
```diff
@@ -39,74 +40,75 @@ class JavaThread;
-      process_self_inner();
-    }
+  bool block_for_operation() {
+    return !_queue.is_empty() || _lock.is_locked();
```
A JavaThread is blocked.
A second thread have just executed a handshake operation for this JavaThread and are on line:
https://github.com/openjdk/jdk/blob/cd784a751a3153939b9284898f370160124ca610/src/hotspot/share/runtime/handshake.cpp#L510
And the queue is empty.

The JavaThread wakes up and changes state from blocked to blocked_trans.
It now checks if it's allowed to complete the transition to e.g. vm.

If a third thread adds to queue before the second thread leaves the loop it's operation can be executed.
But the JavaThread could see the queue as empty. (racey as you say)

The executor takes lock and then checks if the JavaThread is safe for processing.
The JavaThread becomes unsafe and then check if lock is locked.
If the lock is locked we must take slow path to avoid this.

We should also take slow path if there is something on queue to processes.
We are unsafe when we check queue and lock is not held, if we 'miss' that anything is on queue, it's fine.
Since any other thread cannot have seen us as safe and seen the item on queue. (since lock is not held)
Thus not allowed to process the operation.

robehn commented on src/hotspot/share/runtime/handshake.cpp:
```diff
@@ -30,37 +30,44 @@
+};
 
-  bool is_direct() { return _is_direct; }
+class AsyncHandshakeOperation : public HandshakeOperation {
```
This implementation code not part of the interface.
By casting the AsyncHandShakeClosure to a HandshakeClosure before instantiating the HandshakeOperation you can still get is_async() to return true.
And there are a loads of other user error which can be done like stack allocating AsyncHandshakeOperation.
Protecting against all those kinds of errors requires a lot of more code.

robehn commented on src/hotspot/share/runtime/handshake.cpp:
```diff
@@ -30,37 +30,44 @@
+  friend class HandshakeState;
+ protected:
+  HandshakeClosure*   _handshake_cl;
+  int32_t             _pending_threads;
```
Fixed

robehn commented on src/hotspot/share/runtime/handshake.cpp:
```diff
@@ -186,85 +192,42 @@ void VM_Handshake::handle_timeout() {
 
-static void log_handshake_info(jlong start_time_ns, const char* name, int targets, int vmt_executed, const char* extra = NULL) {
-  if (start_time_ns != 0) {
+static void log_handshake_info(jlong start_time_ns, const char* name, int targets, int requester_executed, const char* extra = NULL) {
```
Fixed

robehn commented on src/hotspot/share/runtime/handshake.cpp:
```diff
@@ -373,39 +337,62 @@ bool Handshake::execute_direct(HandshakeClosure* thread_cl, JavaThread* target)
-  _processing_sem(1),
-  _thread_in_process_handshake(false),
-  _active_handshaker(NULL)
+HandshakeState::HandshakeState(JavaThread* thread) :
```
Fixed

robehn commented on src/hotspot/share/runtime/handshake.cpp:
```diff
@@ -414,31 +401,25 @@ void HandshakeState::process_self_inner() {
+    HandshakeOperation* op = pop_for_self();
+    if (op != NULL) {
+      assert(op->_target == NULL || op->_target == Thread::current(), "Wrong thread");
+      assert(_handshakee == Thread::current(), "Wrong thread");
```
Fixed

robehn commented on src/hotspot/share/runtime/handshake.cpp:
```diff
@@ -273,16 +236,16 @@ class VM_HandshakeAllThreads: public VM_Handshake {
-        if (pr == HandshakeState::_success) {
-          handshake_executed_by_vm_thread++;
-        }
+        HandshakeState::ProcessResult pr = thr->handshake_state()->try_process(_op);
```
Yes

sspitsyn commented:
Hi Robbin,
I've looked more at the Serviceability files.
The fix looks good in general. Nice refactoring and simplification with the JvmtiHandshakeClosure.
It is a little unexpected that the do_thread can be executed by a non-JavaThread.
Not sure, what kinds of inconvenience it can cause.
Also, adding the _completed field is somewhat inconsistent.
I'd expect it to be present or not present for almost all JVMTI handshake closures.
I hope, you can comment on why it was added in these particular cases.

robehn commented:
Hi Serguei,

> Hi Robbin,
> I've looked more at the Serviceability files.
> The fix looks good in general. Nice refactoring and simplification with the JvmtiHandshakeClosure.

Good.

> It is a little unexpected that the do_thread can be executed by non-JavaThread's.
> Not sure, what kinds of inconvenience it can cause.

Reading the code I did not find any issues.
Any return values must already be allocated correctly since the executor thread do not wait around.
Thus targeted thread must be the owner of these.
Also extensive testing find no issues.
But as I said to David, we can easily change back to the previous behavior if needed by using the filter mechanism.

> Also, adding the _completed field is somewhat inconsistent.
> I'd expect it to be present or not present for almost all JVMTI handshake closures.
> I hope, you can comment on why it was added in these particular cases.

Two of the handshakes have guarantee checks:
https://github.com/openjdk/jdk/blob/94daf2c7566d882c46cb87b8c28017812bf722ef/src/hotspot/share/prims/jvmtiEnvThreadState.cpp#L326
https://github.com/openjdk/jdk/blob/94daf2c7566d882c46cb87b8c28017812bf722ef/src/hotspot/share/prims/jvmtiEventController.cpp#L345

This is the only placed it is used/needed.
In my previous version I just removed those guarantee because they needed extra code and was suspicious to me.
But it was requested to add them back.

I had quick look now, and from what I can tell it is not guaranteed that we do execute those handshake.
We do not run handshakes on exiting threads (is_terminated()), we set is exiting here:
https://github.com/openjdk/jdk/blob/94daf2c7566d882c46cb87b8c28017812bf722ef/src/hotspot/share/runtime/thread.cpp#L2123
But we remove the JVM TI Thread state way later, here:
https://github.com/openjdk/jdk/blob/94daf2c7566d882c46cb87b8c28017812bf722ef/src/hotspot/share/runtime/thread.cpp#L2207

For example the method ensure_join() which is between set_terminated and removing the JVM TI state can safepoint/handshake.
https://github.com/openjdk/jdk/blob/94daf2c7566d882c46cb87b8c28017812bf722ef/src/hotspot/share/runtime/thread.cpp#L2159

That would trigger the guarantee.

So I believe we should not have those two guarantees and thus the _completed can be removed once again.
I think that should be handled in a separate issue, leaving this as is for now.

robehn commented:
Added comment @dholmes-ora 
Added constructor @dholmes-ora 
Previous renames caused confusing, renamed some methods and moved those not public to private

Running tests

Good to go?

dholmes-ora commented on src/hotspot/share/utilities/filterQueue.hpp:
```diff
@@ -48,24 +52,32 @@ class FilterQueue {
 
-  // MT-Unsafe, external serialization needed.
+  // Applies the match_func to the items in the queue until match_func returns
+  // true and then return false, or there is no more items and then returns
```
Surely "true and then returns true"?
s/there is/there are/

dholmes-ora commented on src/hotspot/share/utilities/filterQueue.hpp:
```diff
@@ -48,24 +52,32 @@ class FilterQueue {
-  // MT-Unsafe, external serialization needed.
+  // Applies the match_func to the items in the queue until match_func returns
+  // true and then return false, or there is no more items and then returns
+  // false. Any pushed item while executing may or may not have match_func
```
s/pushed item/item pushed/

I know what you are trying to say about concurrent pushes but it sounds too non-deterministic - any concurrent push that happens before contains() reaches the end of the queue will have the match_func applied. So in that sense "while executing" only applies to pushes that will be seen; any push not seen had to have happened after execution was complete.

dholmes-ora commented on src/hotspot/share/utilities/filterQueue.hpp:
```diff
@@ -48,24 +52,32 @@ class FilterQueue {
+  // true and then return false, or there is no more items and then returns
+  // false. Any pushed item while executing may or may not have match_func
+  // applied. The method is not re-entrant and must be executed mutually
+  // exclusive other contains and pops calls.
```
// exclusive to other contains() and pop() calls.

dholmes-ora commented on src/hotspot/share/utilities/filterQueue.hpp:
```diff
@@ -48,24 +52,32 @@ class FilterQueue {
 
-  // MT-Unsafe, external serialization needed.
+  // Applies the match_func to all items in the queue returns the item which
+  // match_func return true for and was inserted first. Any pushed item while
```
// Applies the match_func to each item in the queue, in order of insertion, and
// returns the first item for which match_func returns true. Returns false if there are
// no matches or the queue is empty.

robehn replied:
Some offline discussion, fixed.

dholmes-ora commented on src/hotspot/share/utilities/filterQueue.hpp:
```diff
@@ -48,24 +52,32 @@ class FilterQueue {
+  // match_func return true for and was inserted first. Any pushed item while
+  // executing may or may not have be popped, if popped it was the first
+  // inserted match. The method is not re-entrant and must be executed mutual
+  // exclusive with other contains and pops calls.
```
See comments on contains() to ensure pop() and contains() use consistent terminology. Thanks.

dholmes-ora commented on src/hotspot/share/runtime/handshake.cpp:
```diff
@@ -365,27 +376,30 @@ void HandshakeState::add_operation(HandshakeOperation* op) {
 
-static bool processor_filter(HandshakeOperation* op) {
-  return !op->is_asynch();
+static bool non_self_queue_filter(HandshakeOperation* op) {
```
The name "non_self_queue_filter" is really awkward - sorry. But I think we're going to have to revisit the way filtering is named and done once we try to generalise it anyway, so I'll let this pass.

dholmes-ora reviewed (changes requested):
Still some naming issues to resolve and an editing pass on various new comments.

Thanks,
David

robehn reviewed (commented):
Thanks David.

sspitsyn commented:
Robbin, thank you for your answers!
There are JVMTI functions that are specified to return error code JVMTI_ERROR_THREAD_NOT_ALIVE.
As an example, see:
  https://docs.oracle.com/en/java/javase/15/docs/specs/jvmti.html#GetStackTrace
The following functions are impacted by your fix:
  GetOwnedMonitorInfo, GetOwnedMonitorStackDepthInfo, GetCurrentContendedMonitor, GetStackTrace, GetFrameCount, GetFrameLocation, PopFrame

I wonder, how this error code can be ever returned for these functions now.

robehn commented:
> Robbin, thank you for your answers!
> There are JVMTI functions that are specified to return error code JVMTI_ERROR_THREAD_NOT_ALIVE.
> As an example, see:
> https://docs.oracle.com/en/java/javase/15/docs/specs/jvmti.html#GetStackTrace
> The following functions are impacted by your fix:
> GetOwnedMonitorInfo, GetOwnedMonitorStackDepthInfo, GetCurrentContendedMonitor, GetStackTrace, GetFrameCount, GetFrameLocation, PopFrame
> 
> I wonder, how this error code can be ever returned for these functions now.

They should have exactly the same behavior as previously.
All I did was set JVMTI_ERROR_THREAD_NOT_ALIVE as the default value for those handshakes:
https://github.com/openjdk/jdk/blob/3a95750e6474214c62b04fccbfe172d077d3915f/src/hotspot/share/prims/jvmtiEnvBase.hpp#L345
Which simplifies the code.
So only in the two cases where the guarantee's are we can never return that (reset_current_location/enter_interp_only_mode)

If the guarantee's are wrong, the current code have that bug already, so I'm not adding or fixing that.

It should be exactly the same and we have test for at least some of the operations which verifies that the agent gets: JVMTI_ERROR_THREAD_NOT_ALIVE

And this passes t1-8 multiple times, so I'm pretty confident that this does not change any return value.

sspitsyn commented:
Robbin, you are right - thanks.
The JVMTI_ERROR_THREAD_NOT_ALIVE is the default value.

dholmes-ora reviewed (approved):
Hi Robbin,
Thanks for the updates and the slack chat to clarify my misunderstanding of the queuing mechanism.

I agree that the logging statements are somewhat confusing as they were written when the processing logic was much simpler, but I understand now the count of emitted executed operations.

This all looks good to me now.
Thanks,
David

robehn commented:
Thanks all!

/integrate

robehn commented:
/integrate
